File: 3217.java
import java.util.*;

class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        Set<Integer> set = new HashSet<>();
        for (int i : nums) {
            set.add(i);
        }

        while (head != null && set.contains(head.val)) {
            head = head.next;
        }

        if (head == null) {
            return head;
        }

        ListNode cursor = head;
        while (cursor.next != null) {
            if (set.contains(cursor.next.val)) {
                cursor.next = cursor.next.next;
            } else {
                cursor = cursor.next;
            }
        }
        return head;
    }
}

File: Python/3217.py
# Definition for singly-linked list.
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def modifiedList(
        self, nums: List[int], head: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummyHead = ListNode(next=head)
        values = set(nums)
        cursor, prev = head, dummyHead
        while cursor:
            if cursor.val in values:
                prev.next = cursor.next
            else:
                prev = cursor
            cursor = cursor.next
        return dummyHead.next
